이벤트 방출 기능 구현, 보관함 탭 구현

nrurnru opened the pull request:
![screenshot2](https://user-images.githubusercontent.com/53948757/101461320-0a83bb80-397e-11eb-94fa-4b61165b70c5.gif)

![screenshot](https://user-images.githubusercontent.com/53948757/101457173-83801480-3978-11eb-848e-e3892c27b495.gif)

며칠 PR을 안 보냈더니 단위가 너무 커진 것 같네요 죄송합니다😅

- 이벤트 방출 함수 추가
- 보관함 탭 구현
- Reachability 표시 및 구현
- 플레잉바 블러 설정

보관함 탭뷰에서 스크롤뷰를 남발했더니 버그가 너무 많았어서 오래 걸렸네요 ㅜㅜ

nrurnru committed:
[iOS] Feat: 이벤트 발생 기능 구현

뷰에서 이벤트를 발생시키는 동작 구현
코어데이터 이벤트 레이어 분리

nrurnru committed:
[iOS] Feat: 이벤트 서비스 구현, Reachability 구현

코어데이터 함수가 객체를 제네릭으로 받도록 구현
리처빌리티 작성, 기기에서의 테스트 필요

nrurnru committed:
Merge commit '1732afeb032a80c13afd146695775ce95cf3307b' into iOS

# Conflicts:
#	iOS/MiniVIBE/MiniVIBE.xcodeproj/project.pbxproj
#	iOS/MiniVIBE/MiniVIBE/UI/ContentView.swift

nrurnru committed:
[iOS] Feat: 보관함 탭 구현

노래, 아티스트, 앨범, 플레이리스트 탭 구현
각 탭뷰의 높이를 동일하게 맞춰줘야 제대로 동작하는 버그 있음

nrurnru committed:
Merge commit '9569f52428e7735627d8abd8b3f72ca036427dc9' into iOS

# Conflicts:
#	iOS/MiniVIBE/MiniVIBE.xcodeproj/project.pbxproj
#	iOS/MiniVIBE/MiniVIBE/Persistence/Persistence.swift
#	iOS/MiniVIBE/MiniVIBE/Repositories/LocalRepository.swift
#	iOS/MiniVIBE/MiniVIBE/Repositories/ServerRepository.swift
#	iOS/MiniVIBE/MiniVIBE/Services/EventService.swift
#	iOS/MiniVIBE/MiniVIBE/UI/ContentView.swift

nrurnru committed:
[iOS] Feat: 머지 후 이벤트 보내는 방법 수정

Reachability 버그 수정
TabView 높이에 따른 버그 수정
플레잉바 블러 구현

nrurnru committed:
[iOS] Fix: 코어데이터 파일 레퍼런스 문제 수정

nrurnru committed:
[iOS] Fix: 코어데이터 경로 수정

nrurnru committed:
[iOS] Fix: 코어데이터 파일 경로 수정

ChoiDongKyu96 commented on iOS/MiniVIBE/MiniVIBE/Models/CDEvent+CoreDataClass.swift:
```diff
@@ -0,0 +1,15 @@
+
+@objc(CDEvent)
+public class CDEvent: NSManagedObject {
+
```
이 객체는 
@objc가 왜 붙어있으며 의도는 무엇인가요?

nrurnru replied:
이 파일과 아래 파일은 사실 CoreData에서 엔티티를 만들고 클래스 관리를 직접 하기 위해 xCode에서 자동으로 만들어 준 파일입니다. CoreData가 Swift기반이 아닌 오브젝트-C 기반이기 때문에 CoreData가 해당 클래스에 접근하기 위해서는 @objc를 붙여 호환 가능하도록 만들어야 합니다. 이전에 selector 함수를 쓸 때 @objc를 붙였던 것을 생각해 보시면 좋을 것 같네요!

ChoiDongKyu96 commented on iOS/MiniVIBE/MiniVIBE/Models/CDEvent+CoreDataProperties.swift:
```diff
@@ -0,0 +1,28 @@
+        date = event.date
+        tab = Int32(event.tab ?? 0)
+        eventName = event.name
+//        parameter = event.parameter
```
@NSManaged, @nonobjc 프로퍼티 래퍼는 무엇이며 이 주석은 필요한것인가요?

nrurnru replied:
@nonobjc는 위의 경우와 마찬가지로 오브젝트-C 호환이 된 코드(지금으로선 해당 클래스) 안에서 다시 Swift에서만 사용하는 경우 명시적으로 붙여 준다고 합니다. @NSManaged는 해당 멤버들이 entityName에 의해 런타임에 동적으로 결정되어 코어데이터가 구현할 수 있도록 해 준다고 합니다. 아래 주석은 이벤트 파라미터로 사용하기 위해 남겨뒀어요 ㅎㅎ...

ChoiDongKyu96 commented on iOS/MiniVIBE/MiniVIBE/Persistence/Persistence.swift:
```diff
@@ -21,9 +21,8 @@ final class PersistenceController {
-            let request: NSFetchRequest<Event> = Event.fetchRequest()
             let fetchResult = try self.context.fetch(request)
             return fetchResult
         } catch {
```
이 catch문에선 무엇을 해야할까요?

nrurnru replied:
request가 잘못되었을 경우 에러 처리를 해줘야 하겠네요😅 에러타입을 만들어서 다른 에러 부분과 한꺼번에 처리하겠습니다.

ChoiDongKyu96 commented on iOS/MiniVIBE/MiniVIBE/Utilities/Reachability.swift:
```diff
@@ -0,0 +1,28 @@
+    @Published var isConnected: Bool = false
+    
+    init() {
+        monitor.start(queue: DispatchQueue.global())
```
이 경우 글로벌 큐에서 reachability를 계속 observing하고있는건가요?

nrurnru replied:
네 맞습니다. 글로벌 큐에서 계속 비동기로 관찰하다가 isConnected가 바뀔 경우 처리할 수 있도록 했습니다.

ChoiDongKyu96 reviewed (commented):
리쳐블리티와 이벤트까지! 고생많으셨어요 이제 고지가 얼마 남지 않았습니다. 광현님 !! 2주만 힘내보아요😀

Sueaty commented on iOS/MiniVIBE/MiniVIBE/Persistence/Persistence.swift:
```diff
@@ -51,4 +50,12 @@ final class PersistenceController {
         }
     }
+    
+    func newEntity(entityName: String) -> NSManagedObject? {
```
사람마다 방식이 다르긴 하지만 return 해야하는 값이 제대로(?) 있는게 밖으로 나와있는 경우가 더 괜찮지 않나용?ㅋㅋㅋ
guard let 으로 작성하면 else 안에 nil, 바깥에 return NSManagedObject가 오면
읽기  편할 것 같아서요~ㅋㅋㅋ

nrurnru replied:
어 이런 누추한 코드에 귀하신 분이 오셨네요?? ㅋㅋㅋ 이거 똑같은 패턴 이전에 동규님도 말씀하셔서 고쳐놨는데 다들 클린코드의 대가이십니다🥲 리뷰 감사합니다 ㅋㅋ

Sueaty replied:
전 대가 아니고 그냥 대식가🍗🍖🦴🌭🍔

Sueaty commented on iOS/MiniVIBE/MiniVIBE/Repositories/FakeServerRepository.swift:
```diff
@@ -11,14 +11,29 @@ import Combine
+    func send(event: Event) -> AnyPublisher<Void, NetworkError> {
+        let request: RequestProviding
+        
+        if FakeServerRepository.isEnabled {
```
ㅋㅋFakeServer라고 하니까 사기꾼 같쟈나여어어!!ㅋㅋㅋㅋㅋㅋ MockServer 어때요 MockServerㅋㅋ
바꾸긴 늦었나욬ㅋㅋㅋ

nrurnru replied:
아...@ChoiDongKyu96 님의 의견을 들어 보시죠

Sueaty commented on iOS/MiniVIBE/MiniVIBE/Repositories/FakeServerRepository.swift:
```diff
@@ -11,14 +11,29 @@ import Combine
     
-    func send(events: [LocalEvent]) -> AnyPublisher<Void, NetworkError> {
+    func send(event: Event) -> AnyPublisher<Void, NetworkError> {
+        let request: RequestProviding
```
send 함수와 sendAll 함수가 request를 정의하고, 값을 assign 하는 코드가 중복 되는 것 같은데
중복을 줄일 수 있는 방법이 없을까요? 히😋

nrurnru replied:
ㅋㅋㅋ서버동작 확인을 위해서 이벤트를 하나가져와서 넣는거랑 여러개가져와서 넣는 부분인데 조금 더 생각해 보겠습니다... 실제 서버에서도 중복을 줄일 수 있으면 좋겠네요!

Sueaty commented on iOS/MiniVIBE/MiniVIBE/UI/LibraryView/LibraryView.swift:
```diff
@@ -0,0 +1,187 @@
+    var body: some View {
+        NavigationView {
+            ZStack {
+                Color.black.edgesIgnoringSafeArea(.top)
```
`edgesIgnoringSafeArea` 메소드는 deprecated 될 예정이라고 하더라구요!
https://developer.apple.com/documentation/swiftui/image/edgesignoringsafearea(_:) (@TTOzzi 가 알랴줌ㅋㅋㅋㅋ)

`ignoresSafeArea(_:edges:)` 를 써보는건 어때요~?
https://developer.apple.com/documentation/swiftui/image/ignoressafearea(_:edges:)

nrurnru replied:
ㅎㅎㅎㅎ감사합니다...다음부터 이런 기쁜?? 소식이 있다면 부디 @ChoiDongKyu96 님 PR에 써주시면 너무나도 감사할 것 같습니다.😄

Sueaty commented:
ㅋㅋㅋ제 코드 짜기 싫어서 왔냐구여? 어떻게 아셨어요?ㅋㅋㅋㅋㅋㅋㅋㅋㅋ
벌써 네트워크 연결 여부도 구분하시다니... 벌써 코어 데이터 하시다니...
광현님 동규님 진짜 므찌다구여?

nrurnru committed:
[iOS] Fix: 리뷰 반영

do-catch 블록 내에서 리턴 위치 변경
edgesIgnoringSafeArea에서 ignoresSafeArea로 변경
